File: sub_elim.pl
use strict;
use warnings;

my $counter = <STDIN>;

our ($x, $y);

for (1 .. $counter){
	($x, $y) = split /\s/, <STDIN>;
	run ($x, $y);
}

sub getYtype {
	my $y = shift;
	my $char_list;
	for (my $i = 0; $i < length $y; $i++) {
		my $char = substr $y, $i, 1;
		$char_list->{$char} = 1;

		if (keys %$char_list > 2){
			return keys %$char_list;
		}
	}
	return keys %$char_list;
}

sub find_min {
	my $offset = shift;
	my $first = shift;
	my $last = shift;

	my $left = 0;
	my $right = 0;
	while ((substr $x, ($offset - $left), 1) eq $first){		
		$left++;
	}

	while ((substr $x, ($offset + $right + (length $y) - 1), 1) eq $last){
		$right++;
	}

	return [($left >= $right) ? $right : $left, ($left >= $right)];
}

sub find_dup {
	my $offset = shift;
	my $indice = shift;

	my $counter = 0;
	while ((substr $x, ($offset + $counter), 1) eq $indice){
		$counter++;
	}

	return $counter;
}

sub run {
	my $offset = 0;
	my $last_offset;
	my $count = 0;

	my $type = getYtype($y);
	print "type: $type\n";

	while (1) {
		print "index: " . (index $x, $y, $offset) . "\n";
		if ((index $x, $y, $offset) < 0){
			last;
		}else{
			$offset = index $x, $y, $offset;

			if ($type > 2){
				$count++;
				print "count++\n";
				if (defined $last_offset and $offset - $last_offset < length $y and $offset != 0){
					$count--;
					$last_offset = $offset;
					$offset += (length $y) - 1;
					print "conut--\n";
				}else{
					$last_offset = $offset;
				}
				$offset++;
			}elsif ($type == 2){
				my ($min, $direction) = @{find_min($offset, (substr $y, 0, 1), (substr $y, -1, 1))};
				print "sub_min: $min\n";
				print "direction: $direction\n";
				$count += $min;
				if ($direction) {
					$offset += $min;
				}else{
					$offset++;
				}
			}elsif ($type == 1){
				my $dup_length = find_dup($offset, (substr $y, 0, 1));
				print "dup_length: $dup_length\n";
				$count += $dup_length - (length $y) + 1;
				$offset += $dup_length;
			}

			print "offset: $offset\n";
		}
	}

	print "$count\n";

}

File: sub_elim.py
counter = int(raw_input())

global x,y
global char_list
char_list = {}

def getYtype(y):
	for i in range(len(y)):
		char = y[i]
		char_list[char] = 1

		if len(char_list.keys()) > 2 :
			return len(char_list.keys())
		elif (i + 1) < len(y) and y.find(char,i + 1) != (i + 1) and len(char_list.keys()) > 1:
			return 3
	return len(char_list.keys())



def find_min(offset,first,last):
	left = 0
	right = 0
	while(abs(offset-left) < len(x) and x[offset-left] == first):
		left += 1
	
	while(abs(offset + right + len(y) - 1) < len(x) and x[offset + right + len(y) - 1] == last):
		right += 1
	
	if left >= right:
		return right,1
	else:
		return left,0
	
def find_dup(offset, indice):
	counter = 0
	while ((offset + counter) < len(x) and x[offset + counter] == indice):
		counter += 1
	return counter

def recursion_find(sub_x):
	location = sub_x.find(y)
	if location < 0:
		return 0 

	try1 = 1 + recursion_find(sub_x[0:location] + sub_x[(location + 1):]) 
		
	if (location + len(y) >= len(sub_x)):
		return 1
	if not sub_x[location + len(y)] in char_list.keys():
		return 1
	try2 = 1 + recursion_find(sub_x[0:(location + len(y))] + (sub_x[(location + len(y) + 1):]))

	if (try1 >= try2):
		return try2
	else:
		return try1

def run():
	offset = 0
	last_offset = -1
	count = 0

	type = getYtype(y)

	while(True):

		if (x.find(y,offset) < 0):
			break
		else:
			offset = x.find(y,offset)

			if type > 2:
				count += 1
				if last_offset != -1 and offset - last_offset < len(y) and offset != 0:
					count -= 1
					last_offset = offset
					offset += len(y) - 1
				else:
					last_offset = offset
				offset += 1
			elif type == 2:
				count = recursion_find(x)
				break
			elif type == 1:
				dup_length = find_dup(offset, y[0])
				count += dup_length - len(y) + 1;
				offset += dup_length;


	print(count)

for i in range(counter):
	x,y = raw_input().split()
	run()

